src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,638 @@
+            F::fill_with_bitvector_representation(
+                entry,
+                &mut encoded[l * self.bits_per_entry..(l + 1) * self.bits_per_entry],
+            )?;
+        }
+
+        // (II) Vector norm.
+        // Compute the norm of the input vector.
+        let field_entries = integer_entries.iter().map(|&x| F::from(x));
+        let norm =
+            compute_norm_of_entries(field_entries, self.bits_per_entry, F::one(), &mut |x| {
+                Ok(x * x)
+            })?;
+        let norm_int = <F as FieldElement>::Integer::from(norm);
+
+        // Write the norm into the `entries` vector.
+        F::fill_with_bitvector_representation(
+            &norm_int,
+            &mut encoded[self.range_norm_begin..self.range_norm_end],
+        )?;
+
+        Ok(encoded)
+    }
+
+    fn decode_result(
+        &self,
+        data: &[F],
+        num_measurements: usize,
+    ) -> Result<Vec<<T as CompatibleFloat<F>>::Float>, FlpError> {
+        if data.len() != self.entries {
+            return Err(FlpError::Decode("unexpected input length".into()));
+        }
+        let mut res = vec![];
+        for d in data {
+            let decoded = <T as CompatibleFloat<F>>::to_float(*d, num_measurements);
+            res.push(decoded);
+        }
+        Ok(res)
+    }
+
+    fn gadget(&self) -> Vec<Box<dyn Gadget<F>>> {
+        // This gadgets checks that a field element is zero or one.
+        // It is called for all the "bits" of the encoded entries
+        // and of the encoded norm.
+        let gadget0 = PolyEval::new(
+            self.range_01_checker.clone(),
+            self.bits_per_entry * self.entries + self.bits_for_norm,
+        );
+
+        // This gadget computes the square of a field element.
+        // It is called on each entry during norm computation.
+        let gadget1 = PolyEval::new(self.square_computer.clone(), self.entries);
+
+        let res: Vec<Box<dyn Gadget<F>>> = vec![Box::new(gadget0), Box::new(gadget1)];
+        res
+    }
+
+    fn valid(
+        &self,
+        g: &mut Vec<Box<dyn Gadget<F>>>,
+        input: &[F],
+        joint_rand: &[F],
+        num_shares: usize,
+    ) -> Result<F, FlpError> {
+        self.valid_call_check(input, joint_rand)?;
+
+        // Ensure that all submitted field elements are either 0 or 1.
+        // This is done for:
+        //  - all vector entries (each of them encoded in `self.bits_per_entry`
+        //    field elements)
+        //  - the submitted norm (encoded in `self.bits_for_norm` field
+        //    elements)
+        //
+        // Since all input vector entry (field-)bits, as well as the norm bits,
+        // are contiguous, we do the check directly for all bits from 0 to
+        // entries*bits_per_entry + bits_for_norm.
+        //
+        // Check that each element is a 0 or 1:
+        let range_check =
+            call_gadget_on_vec_entries(&mut g[0], &input[0..self.range_norm_end], joint_rand[0])?;
```

Comment:
The gadget has degree 2 and the number of gadget calls is `entries * n + 2 * n  - 2`, Thus the length of the proof polynomial is 

```
2 * (next_power_of_2(1 + entries*n + 2*n - 2) - 1) + 1 
= 2 * (next_power_of_2(entries*n + 2*n - 1) - 1) + 1
<= 2 * (2 * (entries*n + 2*n - 1) - 1) + 1 // since next_power_of_2(x) <= 2*x for all x
= 2 * (entries*2*n + 4*n - 2 - 1) + 1
= entries*4*n + 8*n - 1
```

Does that look right?

It should be possible to shorten the proof to roughly `O(sqrt(entries*4*n + 8*n - 1))` using the `ParallelSum` gadget. The downside is that proof generation/verification is more expensive. Still, the trade-off might be worth it here. For a working example of this, see the `CountVec` type.

Before going much further, I suggest we benchmark this code as described here: https://github.com/divviup/libprio-rs/pull/283#pullrequestreview-1082914192. We need to get a sense of how long proof generation takes and how long the Prio3 input shares are. We could then measure the concrete cost, in terms of computation, of shortening the proof.

Reply:
Thanks @MxmUrw and @ooovi for adding a benchmark. Here's what I get on my laptop:
```
prio3 fixedpoint16 boundedl2 vec (1000 entries) size = 406864
Benchmarking prio3 fixedpoint16 boundedl2 vec (1000 entries): Collecting 100 samples in estimated 6.0199 s (300                                                                                                               prio3 fixedpoint16 boundedl2 vec (1000 entries)
                        time:   [19.945 ms 20.012 ms 20.080 ms]
                        change: [-29.655% -29.125% -28.655%] (p = 0.00 < 0.05)
                        Performance has improved.
```

The client runtime (20 ms) is definitely on the high end for Prio3 types, but I think reasonable for many applications. I'm more worried about the size of the input shares (406,864 bytes, or roughly 397KB). Thus I think it would be worth considering optimizations to the circuit that would drive down the size of the proof.

I would strongly recommend we use the `ParallelSum` gadget. I put together a change to demonstrate how this would be used for the range check part of the circuit: https://github.com/divviup/libprio-rs/commit/21491ac702ab06739f57dedea915b3b54ee76815.

```
prio3 fixedpoint16 boundedl2 vec (1000 entries) size = 152864
Benchmarking prio3 fixedpoint16 boundedl2 vec (1000 entries): 
[...]
prio3 fixedpoint16 boundedl2 vec (1000 entries)
                        time:   [64.192 ms 64.988 ms 65.953 ms]
                        change: [+220.42% +224.74% +230.01%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 14 outliers among 100 measurements (14.00%)
  4 (4.00%) high mild
  10 (10.00%) high severe
```

The headline here is that the input shares are now just 149KB. The downside is that the client runtime is tripled, due to the more expensive (i.e., higher arithmetic degree) circuit (see the BlindPolyEval gadget). However:
1. By multithreading the circuit evaluation, we can make this competitive with a lower degree circuit. See https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/83.
2. As the proof shrinks, proof generation time tends to improve. See https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/83.

We can use a similar trick for the second gadget as well, driving down the proof size even further. However, for this gadget we wouldn't even need the more expensive BlindPolyEval gadget, since all we want to compute is the sum of the squares.